File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/agent/AgentObservers.java
package org.hyperledger.indy.sdk.agent;

/**
 * agent.rs observers
 */
/**
 * Observer interfaces for agents.
 */
public final class AgentObservers {

	private AgentObservers() {

	}

	/**
	 * An observer to receive notification of the establishment of a connection.
	 */
	public interface ConnectionObserver {

		/**
		 * Called when a connection is established.
		 * 
		 * @param listener The listener the connection is associated with.  Can be null for outgoing connections.
		 * @param connection The connection that has been established.
		 * @param senderDid The DID of the identity that initated the connection.
		 * @param receiverDid The DID of the identity that received the connection.
		 * @return A MessageObserver that will receive notifications when the connection receives a message.
		 */
		public MessageObserver onConnection(Agent.Listener listener, Agent.Connection connection, String senderDid, String receiverDid);
	}

	/**
	 * An observer to receive notification of the receipt of a message.
	 */
	public interface MessageObserver {

		/**
		 * Called when a message is received.
		 * 
		 * @param connection The connection the message was received on.
		 * @param message The received message.
		 */
		public void onMessage(Agent.Connection connection, String message);
	}
}

File: wrappers/dotnet/indy-sdk-dotnet/Wrapper/AgentObservers.cs
﻿namespace Indy.Sdk.Dotnet.Wrapper
{
    /// <summary>
    /// Observer classes for use with the Agent API.
    /// </summary>
    public sealed class AgentObservers
    {
        /// <summary>
        /// Prevents class being instantiated.
        /// </summary>
        private AgentObservers()
        {
        }

        /// <summary>
        /// Interface for handling the event raised on establishment of a connection.
        /// </summary>
        public interface ConnectionObserver
        {
            /// <summary>
            /// Called when a connection is established by a remote party.
            /// </summary>
            /// <param name="listener">The listener the connection was established on.</param>
            /// <param name="connection">The connection that was created.</param>
            /// <param name="senderDid">The DID of the sender.</param>
            /// <param name="receiverDid">The DID of the receiver.</param>
            /// <returns>A MessageObserver that will receive notifications when messages are received on the connection.</returns>
            MessageObserver OnConnection(Agent.Listener listener, Agent.Connection connection, string senderDid, string receiverDid);
        }

        /// <summary>
        /// Interface for handling the event raised when a message is received.
        /// </summary>
        public interface MessageObserver
        {
            /// <summary>
            /// Called when a message is received.
            /// </summary>
            /// <param name="connection">The connection the message was received on.</param>
            /// <param name="message">The content of the received message.</param>
            void OnMessage(Agent.Connection connection, string message);
        }
    }
}
